File: c++/source/p007.cpp
#include <iostream>
#include <math.h>

using namespace std;

bool isPrime(int n);

int main()  {
    
    int count = 0;
    int prime = 1;

    while (count < 10001) {
        prime ++;
        if (isPrime(prime)) {
            count ++;
        }
    }

    cout << prime << endl; 

}


bool isPrime(int n) {
    if (n <= 1) {
        return false;
    } else if (n <= 3) {
        return true;
    } else if (n % 2 == 0) {
        return false;
    } else {
        for (int i = 3; i <= sqrt(n); i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        return true;
    }
}

File: python/p007.py
import euler


def calculate():

    # Old Code: 7.7s
    # n = 10001
    # primes = [2]
    # attempt = 3
    # while len(primes) < n:
    #     if all(attempt % prime != 0 for prime in primes):
    #         primes.append(attempt)
    #     attempt += 2
    # return str(primes[-1])

    # New Code: 0.1s
    def find_prime(n):
        count = 0
        prime = 1

        while count < n:
            prime += 1
            if euler.is_prime(prime):
                count += 1
        return prime

    return str(find_prime(10001))


if __name__ == "__main__":
    print(calculate())
